File: src/chroma-cloud/code/python/error-handling.py
# @snippet:imports
import json
import os
import time

import chromadb
from chromadb.errors import NotFoundError
# @end


def get_cloud_config() -> dict[str, str]:
    api_key = os.environ["CHROMA_API_KEY"]
    tenant = os.environ["CHROMA_TENANT"]
    database = os.environ["CHROMA_DATABASE"]
    return {"api_key": api_key, "tenant": tenant, "database": database}


# @snippet:connection-errors
def connect_with_retry(max_retries: int = 3) -> chromadb.ClientAPI:
    """Connect to Chroma Cloud with exponential backoff retry."""
    client = chromadb.CloudClient(**get_cloud_config())

    for attempt in range(1, max_retries + 1):
        try:
            client.heartbeat()
            return client
        except Exception as e:
            if attempt == max_retries:
                raise ConnectionError(
                    f"Failed to connect to Chroma Cloud after {max_retries} attempts: {e}"
                )

            time.sleep(2 ** (attempt - 1))

    raise ConnectionError("Unreachable")


# @end

# @snippet:collection-not-found
client = chromadb.CloudClient(**get_cloud_config())

try:
    collection = client.get_collection(name="my_collection")
except NotFoundError:
    print("Collection not found, creating it...")
    collection = client.get_or_create_collection(name="my_collection")
# @end

# @snippet:safe-collection-access
client = chromadb.CloudClient(**get_cloud_config())
collection = client.get_or_create_collection(name="my_collection")

results = collection.query(
    query_texts=["search query"],
    n_results=5,
)

if results["documents"] and len(results["documents"][0]) > 0:
    first_doc = results["documents"][0][0]
else:
    pass
# @end

# @snippet:validation-errors
client = chromadb.CloudClient(**get_cloud_config())


def validate_document(doc: str) -> bool:
    byte_size = len(doc.encode("utf-8"))
    return byte_size <= 16384


def validate_metadata(metadata: dict) -> bool:
    if len(metadata.keys()) > 32:
        return False

    json_size = len(json.dumps(metadata).encode("utf-8"))
    return json_size <= 4096


def safe_add(
    collection_name: str,
    ids: list[str],
    documents: list[str],
    metadatas: list[dict] | None = None,
) -> None:
    for doc in documents:
        if not validate_document(doc):
            raise ValueError("Document exceeds 16KB limit")

    if metadatas:
        for meta in metadatas:
            if not validate_metadata(meta):
                raise ValueError("Metadata exceeds limits (4KB or 32 keys)")

    collection = client.get_or_create_collection(name=collection_name)
    collection.add(ids=ids, documents=documents, metadatas=metadatas)


# @end

# @snippet:batch-operations
client = chromadb.CloudClient(**get_cloud_config())


def batch_add(
    collection_name: str,
    ids: list[str],
    documents: list[str],
    batch_size: int = 100,
) -> dict:
    collection = client.get_or_create_collection(name=collection_name)
    failures = []

    for i in range(0, len(ids), batch_size):
        batch_ids = ids[i : i + batch_size]
        batch_docs = documents[i : i + batch_size]

        try:
            collection.add(ids=batch_ids, documents=batch_docs)
        except Exception as e:
            failures.append({"index": i, "error": str(e)})

    total_batches = (len(ids) + batch_size - 1) // batch_size
    return {"total_batches": total_batches, "failures": failures}


# @end

# @snippet:cloud-errors
def create_cloud_client() -> chromadb.ClientAPI:
    """Create CloudClient with error handling."""
    client = chromadb.CloudClient(**get_cloud_config())

    try:
        client.heartbeat()
        return client
    except Exception as e:
        error_msg = str(e).lower()
        if "401" in error_msg or "unauthorized" in error_msg:
            raise PermissionError("Invalid or expired API key") from e
        if "404" in error_msg or "not found" in error_msg:
            raise NotFoundError(
                "Tenant or database not found - check configuration"
            ) from e
        if "429" in error_msg or "rate" in error_msg:
            raise RuntimeError("Rate limit exceeded - implement backoff") from e
        raise


# @end

File: src/chroma-cloud/code/typescript/error-handling.ts
// @snippet:imports
import { CloudClient } from 'chromadb';
import { DefaultEmbeddingFunction } from '@chroma-core/default-embed';
// @end

function getCloudConfig() {
  const apiKey = process.env.CHROMA_API_KEY;
  const tenant = process.env.CHROMA_TENANT;
  const database = process.env.CHROMA_DATABASE;

  const missing: string[] = [];
  if (!apiKey) missing.push('CHROMA_API_KEY');
  if (!tenant) missing.push('CHROMA_TENANT');
  if (!database) missing.push('CHROMA_DATABASE');

  if (missing.length > 0) {
    throw new Error(
      `Missing required environment variables: ${missing.join(', ')}`
    );
  }

  return { apiKey, tenant, database };
}

// @snippet:connection-errors
async function connectWithRetry(maxRetries = 3): Promise<CloudClient> {
  const config = getCloudConfig();
  const client = new CloudClient(config);

  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      await client.heartbeat();
      return client;
    } catch (error) {
      if (attempt === maxRetries) {
        throw new Error(
          `Failed to connect to Chroma Cloud after ${maxRetries} attempts: ${error}`
        );
      }

      const delay = Math.pow(2, attempt - 1) * 1000;
      await new Promise((resolve) => setTimeout(resolve, delay));
    }
  }

  throw new Error('Unreachable');
}
// @end

// @snippet:collection-not-found
const config = getCloudConfig();
const client = new CloudClient(config);
const embeddingFunction = new DefaultEmbeddingFunction();

try {
  await client.getCollection({
    name: 'my_collection',
    embeddingFunction,
  });
} catch (error) {
  if (error instanceof Error && error.message.includes('does not exist')) {
    console.log('Collection not found, creating it...');
    await client.getOrCreateCollection({
      name: 'my_collection',
      embeddingFunction,
    });
  } else {
    throw error;
  }
}
// @end

// @snippet:safe-collection-access
const config2 = getCloudConfig();
const client2 = new CloudClient(config2);
const embeddingFunction2 = new DefaultEmbeddingFunction();

const collection = await client2.getOrCreateCollection({
  name: 'my_collection',
  embeddingFunction: embeddingFunction2,
});

const results = await collection.query({
  queryTexts: ['search query'],
  nResults: 5,
});

if (results.documents[0] && results.documents[0].length > 0) {
  const firstDoc = results.documents[0][0];
} else {
  // No results found
}
// @end

// @snippet:validation-errors
const config3 = getCloudConfig();
const client3 = new CloudClient(config3);
const embeddingFunction3 = new DefaultEmbeddingFunction();

function validateDocument(doc: string): boolean {
  const byteSize = new TextEncoder().encode(doc).length;
  return byteSize <= 16384;
}

function validateMetadata(
  metadata: Record<string, string | number | boolean>
): boolean {
  const keys = Object.keys(metadata);
  if (keys.length > 32) return false;

  const jsonSize = new TextEncoder().encode(JSON.stringify(metadata)).length;
  return jsonSize <= 4096;
}

async function safeAdd(
  collectionName: string,
  ids: string[],
  documents: string[],
  metadatas?: Record<string, string | number | boolean>[]
) {
  for (const doc of documents) {
    if (!validateDocument(doc)) {
      throw new Error('Document exceeds 16KB limit');
    }
  }

  if (metadatas) {
    for (const meta of metadatas) {
      if (!validateMetadata(meta)) {
        throw new Error('Metadata exceeds limits (4KB or 32 keys)');
      }
    }
  }

  const collection = await client3.getOrCreateCollection({
    name: collectionName,
    embeddingFunction: embeddingFunction3,
  });

  await collection.add({ ids, documents, metadatas });
}
// @end

// @snippet:batch-operations
const config4 = getCloudConfig();
const client4 = new CloudClient(config4);
const embeddingFunction4 = new DefaultEmbeddingFunction();

async function batchAdd(
  collectionName: string,
  ids: string[],
  documents: string[],
  batchSize = 100
) {
  const collection = await client4.getOrCreateCollection({
    name: collectionName,
    embeddingFunction: embeddingFunction4,
  });

  const failures: { index: number; error: string }[] = [];

  for (let i = 0; i < ids.length; i += batchSize) {
    const batchIds = ids.slice(i, i + batchSize);
    const batchDocs = documents.slice(i, i + batchSize);

    try {
      await collection.add({
        ids: batchIds,
        documents: batchDocs,
      });
    } catch (error) {
      failures.push({
        index: i,
        error: error instanceof Error ? error.message : String(error),
      });
    }
  }

  return { totalBatches: Math.ceil(ids.length / batchSize), failures };
}
// @end

// @snippet:cloud-errors
async function createCloudClient() {
  const config = getCloudConfig();
  const client = new CloudClient(config);

  try {
    await client.heartbeat();
    return client;
  } catch (error) {
    if (error instanceof Error) {
      if (
        error.message.includes('401') ||
        error.message.includes('Unauthorized')
      ) {
        throw new Error('Invalid or expired API key');
      }
      if (
        error.message.includes('404') ||
        error.message.includes('not found')
      ) {
        throw new Error('Tenant or database not found - check configuration');
      }
      if (error.message.includes('429') || error.message.includes('rate')) {
        throw new Error('Rate limit exceeded - implement backoff');
      }
    }

    throw error;
  }
}
// @end
